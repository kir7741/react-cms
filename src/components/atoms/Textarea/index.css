.textarea-wrapper {
	position: relative;
	display: grid;
	margin-bottom: 10px;
	padding-bottom: 1.25em;
	font-size: 16px;

	& span {
		position: absolute;
		bottom: 0px;
		font-size: 14px;
		color: var(--error-red);

	}

}

.textarea {
	padding: 10px 15px;
	border: 1px solid var(--primary-gray);
	border-radius: 6px;

	&.error {
		border-color: var(--error-red);
	}

	&:focus {
    outline: none;
	}

}
