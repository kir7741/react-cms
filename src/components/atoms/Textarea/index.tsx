import React, { TextareaHTMLAttributes, useEffect, useRef } from 'react';
import classnames from 'classnames';

import styles from './index.css';


export interface TextareaProperty extends TextareaHTMLAttributes<HTMLTextAreaElement> {

	/**
	 *  class 名稱
	 *
	 * @type {string}
	 * @memberof TextareaProperty
	 */
	className?: string;

	/**
	 * 錯誤訊息
	 *
	 * @type {string}
	 * @memberof TextareaProperty
	 */
	errorMsg?: string;

	/**
	 * 是否在 blur 時觸發檢核
	 *
	 * @type {boolean}
	 * @memberof TextareaProperty
	 */
	 validOnBlur?: boolean;

	 /**
		* 更新表單檢核狀態
		*
		* @memberof TextareaProperty
		*/
	 updateCtrlValidity?: () => void,

	 /**
		* change 時觸發的函式
		*
		* @memberof TextareaProperty
		*/
	 onChangeValue: (val: string | number) => void

	 /**
		* blur 時觸發的函式
		*
		* @memberof TextareaProperty
		*/
	 blur?: () => void

}

const Textarea: React.FC<TextareaProperty> = ({
	className,
	placeholder = '',
	value = '',
	errorMsg = '',
	validOnBlur = false,
	disabled,
	cols = 10,
	rows = 10,
	updateCtrlValidity = () => {},
	onChangeValue,
	blur = () => {}
}) => {

	const inInit = useRef(true);

	const handleInputChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
		e.preventDefault();
		const val = e.target.value;
		onChangeValue(val);
	}

	useEffect(() => {

		if (inInit.current) {
			inInit.current = false;
			return;
		}

		if (!validOnBlur) {
			updateCtrlValidity();
		}

	}, [value])

	return (
		<div className={classnames(styles.textareaWrapper, className)}>
			<textarea
				cols={cols}
				rows={rows}
				disabled={disabled}
				className={classnames(styles.textarea, errorMsg && styles.error)}
				value={value}
				placeholder={placeholder}
				onBlur={() => {
					if (validOnBlur) {
						updateCtrlValidity();
					}
					blur();
				}}
				onChange={event => handleInputChange(event)}
			/>
			{errorMsg && <span>{errorMsg}</span>}
		</div>
	);

}

export default Textarea;
