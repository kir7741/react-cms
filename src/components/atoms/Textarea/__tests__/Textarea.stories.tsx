import React from 'react';
import { storiesOf } from '@storybook/react';

import Textarea from 'components/atoms/Textarea';

const stories = storiesOf('atoms/Textarea', module);

stories.add('__interactive', () => <Textarea />);
