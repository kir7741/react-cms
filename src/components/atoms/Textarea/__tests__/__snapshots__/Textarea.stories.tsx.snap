// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="textareaWrapper"
  >
    <textarea
      class="textarea"
      cols="10"
      placeholder=""
      rows="10"
    />
  </div>
</DocumentFragment>
`;
